pkg/sds/server/v2/reconciler.go

```diff
@@ -74,103 +83,116 @@ func (d *DataplaneReconciler) Reconcile(dataplaneId core_model.ResourceKey) erro
 		return nil
 	}
 
-	generateSnapshot, reason, err := d.shouldGenerateSnapshot(proxyID, mesh)
+	generateSnapshot, reason, err := d.shouldGenerateSnapshot(proxyID, mesh, dataplane)
 	if err != nil {
 		return err
 	}
 
 	if generateSnapshot {
 		sdsServerLog.Info("Generating the Snapshot.", "dataplaneId", dataplaneId, "reason", reason)
-		snapshot, err := d.generateSnapshot(dataplane, mesh)
+		snapshot, info, err := d.generateSnapshot(dataplane, mesh)
 		if err != nil {
 			return err
 		}
-		d.sdsMetrics.CertGenerations(envoy_common.APIV2).Inc()
-		if err := d.updateInsights(dataplaneId, snapshot); err != nil {
-			// do not stop updating Envoy even if insights update fails
-			sdsServerLog.Error(err, "Could not update Dataplane Insights", "dataplaneId", dataplaneId)
-		}
+
+		d.sdsMetrics.CertGenerations(envoy_common.APIV3).Inc()
 		if err := d.cache.SetSnapshot(proxyID, snapshot); err != nil {
 			return err
 		}
+		d.setSnapshotInfo(proxyID, info)
```

Comment:
Probably `d.cache.SetSnapshot` and `d.setSnapshotInfo` should be atomic. Otherwise concurrent `Reconcile` call will run `d.shouldGenerateSnapshot` where we check if snapshot exists and then we call `d.snapshotInfo`. So the following execution is possible:
```
goroutine-1                       goroutine-2
---------------------------------------------------------
d.cache.SetSnapshot
                                  shouldGenerateSnapshot
                                  d.cache.GetSnapshot
                                  info := d.snapshotInfo # info will be empty
d.setSnapshotInfo
```
It's not super dangerous, I think in the worst scenario we'll just regenerate certificates twice, but probably it's better to fix it

Reply:
for one dataplane there is only one watchdog with reconciling so the following scenario should not ever happen.